packages/dds/tree/docs/wip/schema-evolution.md

```diff
@@ -0,0 +1,583 @@
+Maintain view schema B in application source code : 2, 1d
+Documents contain stored schema B                 : 2, 1d
+```
+
+Unfortunately, many applications operate under additional constraints.
+For example, what if the developer does not have control over all the documents?
+Perhaps the documents can be stored locally on a user's computer, or can be hosted by another service that is independent of the developer.
+What if the developer does not have control over when a client updates their application?
+Even "live service" applications, like web apps, might allow users to continue using an older/cached version of an application client long after a deployment of a new version.
+In these scenarios, when does the document get upgraded?
+When does the client move to the new version of the application?
+The developer cannot know, and therefore cannot guarantee that the stored schema and view schema of their application will be the same for all clients and documents.
+
+This is problematic.
+If view schema B is deployed to the application, a client might encounter a document that is still using stored schema A (i.e. the view schema is ahead of the stored schema).
+Or, a client might avoid updating even after its document has upgraded to the stored schema B (i.e. the view schema is behind the stored schema).
+The application cannot function unless the view schema and the stored schema agree.
+
+How then does a developer in this situation go about upgrading their application schema?
+The developer must do a **staged rollout** - an incremental set of application deployments that release client code only when it is safe to do so.
+
+1. Initially, the documents are using stored schema A and the application is written against view schema A.
+2. View schema B is written into the application client code.
+   The application's UI, etc. is written against it, but the existing code that (written against view schema A) continues to be maintained.
+   Thus the application logic becomes bifurcated according to the schema.
+   If the application loads a document with stored schema A, it views that document with view schema A.
+   If instead it loads a document with stored schema B, it views it with view schema B.
+   Both versions of the view/UI/application/etc. must continue to be maintained until after step 4.
+   This new version of the application is deployed and clients can begin updating to it.
+3. The developer waits until all (or most) clients have upgraded to this new application version.
+   The amount of clients that have moved to the new version is known as the **client saturation**.
+   It may be impossible to know for sure when 100% client saturation has been achieved, so a developer may estimate when the client saturation is sufficient - e.g. after 99% of online clients in the last 30 days had updated the new version.
+   After client saturation, the stored schema of the documents can now be upgraded to B.
```

Comment:
How does it happen? Is it a full rewrite of a document?
What happens with trailing ops?
What happens if two changes of schema crisscross?

This is the most interesting part that deserves a lot of details (maybe not in this document, but that's something that will define success of migration).

Similar questions exist even when using adapters, including bidirectional adapters / flows.

Let's use classical example - adding a new field. What happens if two clients add the same field? In most cases clients would be Ok with FWW semantics here (LWW could result in data loss). I can go into more details on scenarios (from Loop) and reasoning.